### ISSUE TITLE
Noticeable Scene Load increase in 2.0 release compared to beta

### ISSUE THREAD

**Issue by KXI-System** [YarnSpinnerTool/YarnSpinner-Unity#143]
**What is the current behavior?**
There is a noticeable increase in scene loading times found between beta5 and 2.0.1, with the loading times increasing even more with registered functions and commands though noticeable with none.

**Please provide the steps to reproduce, and if possible a minimal demo of the problem**:
The following is tested with a new blank Unity project in 2021.1.6f1, with the following yarn file:
```
title: Start
tags: 
position: 100,100
---
Chester: Hello World! #line:08a14c9
===
```
I created a new yarn project from this file and put in the Dialogue Prefab in the scene. Modifying no other settings on the prefab other than hooking up the test yarn project. I proceeded to time and record how long it took for my system to enter the game after pressing play (with literally a stopwatch in hand), I did the same setup and test in beta5 and 2.0.1 installing from Git.

Results averaged between three tests:
beta5: ~3.79 seconds
2.0.1: ~9.30 seconds

There might be human error involved with the timing, but there is a noticeable increase with no changes. Also of note; Unity's loading window appeared during these tests on 2.0.1 (holding on at Application.EnterPlayMode the longest) but not in beta5. 
There was a point brought up on the Discord Server with the examples having different loading times in 2.0.1, but in my tests with the Intro sample showed no meaningful difference between it and my empty test scene.

EDIT: I forgot to mention that my machine is fairly beefy (but not top of the line), though the specs of my machine aren't the focus but rather the difference between the load times. I imagine on bigger projects or less capable hardware that this issue is much bigger.

**Comment by KXI-System** [comment]
Some more additional testing, when hitting play with an empty scene that does not have the dialogue prefab:

- There is only a ~0.5s-1s difference between beta5 and 2.0.1
- Loading into the test scene from the empty scene (using single load scene mode) doesn't have any major hiccups or difference in load times between the two versions

To clarify, entering into an empty scene in beta5 took ~3.65 but in 2.0.1 took ~4.17. Despite that, both loads into the scene containing the exact same setup with the Dialogue prefab basically instantly from the empty scene.

**Comment by KXI-System** [comment]
More testing with registered functions through code and attributes, both setups have a script registering 30 commands and functions through both code and attributes.
I couldn't find any discernable difference between load times compared to not registering any commands, though 2.0.1 still takes a much longer time to enter play mode compared to beta5 as usual. Though all of the commands and functions are very simple like so:
```csharp
dialogueRunner.AddCommandHandler("TestCmd", () => { });
dialogueRunner.AddFunction("TestFunc", () => { return ""; });
dialogueRunner.AddCommandHandler("TestIECmd", () => { return StartCoroutine(IE()); });
...
private IEnumerator IE() { yield return null; }
```

**Comment by desplesda** [comment]
This issue turns out to be caused by the way that `ActionManager` searches assemblies for methods tagged with `YarnCommand` and `YarnFunction`. As of 2.0.1, we're searching _every_ method in _every_ assembly, when we don't need to (for example, we're currently searching assemblies like System.dll and UnityEngine.dll, which contain a large number of methods and never contain Yarn commands.) This incurred a significant loading time, and also meant that the garbage collector was allocating a lot of memory during the search (in some of my measurements, it was in the low hundreds of MB. This is very bad for mobile platforms!)

What I've done is:

1. Made `ActionManager` search `Assembly-CSharp` (the default assembly that Unity puts C# code into), along with specifically named assemblies upon request. 
2. `DialogueRunner` makes this request when loading a `YarnProject`; I've added a new property, `searchAssembliesForActions`, which contains the names of assemblies to load.
3. Updated `YarnProjectImporter` to store references to Assembly Definition Assets, which are how users define assemblies in Unity. When the `YarnProject` is imported, `searchAssembliesForActions` is populated. By default, _all_ assembly definitions are considered; the user may opt to only include a specific subset. 

As a result, there's now an additional entry in the Inspector for `.yarnproject` assets:

![Specifying assemblies in a Yarn Project](https://user-images.githubusercontent.com/901768/147647760-616133c9-914b-4436-9665-dafe7b04432f.png)


This strikes me as the best tradeoff between ensuring that the user is able to make use of their commands and functions when they're defined in code, and reducing the startup burden.

I've implemented this in db2177cf8c303583382e043afbe12e7f0dc97df3, and it'll get released as part of v2.0.2. Thanks to @KXI-System for their detailed reports, and to other users on Discord for reporting this and contributing additional info that helped track it down!
